File: cpp/tests/test_optimize.cpp
// Tests for the native Nelder-Mead minimiser.

#include "irm/optimize.hpp"

#include <cmath>
#include <stdexcept>
#include <vector>

#include <gtest/gtest.h>

namespace {

TEST(NelderMead, MinimisesShiftedQuadratic) {
    const auto f = [](const std::vector<double>& p) {
        const double dx = p[0] - 1.5, dy = p[1] + 2.0;
        return dx * dx + 3.0 * dy * dy;
    };
    const irm::NelderMeadResult res = irm::nelder_mead(f, {0.0, 0.0});
    EXPECT_TRUE(res.converged);
    EXPECT_NEAR(res.x[0], 1.5, 1e-7);
    EXPECT_NEAR(res.x[1], -2.0, 1e-7);
    EXPECT_NEAR(res.fx, 0.0, 1e-12);
}

TEST(NelderMead, MinimisesRosenbrock) {
    const auto f = [](const std::vector<double>& p) {
        const double a = 1.0 - p[0];
        const double b = p[1] - p[0] * p[0];
        return a * a + 100.0 * b * b;
    };
    const irm::NelderMeadResult res = irm::nelder_mead(f, {-1.2, 1.0});
    EXPECT_TRUE(res.converged);
    EXPECT_NEAR(res.x[0], 1.0, 1e-6);
    EXPECT_NEAR(res.x[1], 1.0, 1e-6);
}

TEST(NelderMead, NonConvergenceReportedNotThrown) {
    const auto f = [](const std::vector<double>& p) {
        const double a = 1.0 - p[0];
        const double b = p[1] - p[0] * p[0];
        return a * a + 100.0 * b * b;
    };
    const irm::NelderMeadResult res =
        irm::nelder_mead(f, {-1.2, 1.0}, 0.1, 1e-10, 1e-14, /*maxiter=*/5);
    EXPECT_FALSE(res.converged);
    EXPECT_EQ(res.iterations, 5);
}

TEST(NelderMead, InvalidStartThrows) {
    const auto f = [](const std::vector<double>&) { return 0.0; };
    EXPECT_THROW(irm::nelder_mead(f, {}), std::invalid_argument);
    EXPECT_THROW(irm::nelder_mead(f, {std::nan("")}), std::invalid_argument);
}

}  // namespace

File: python/tests/test_optimize.py
"""Native Nelder-Mead, cross-checked against scipy (tests only)."""

from __future__ import annotations

import pytest
import scipy.optimize as sopt

from irm import nelder_mead


def _rosenbrock(p) -> float:
    x, y = p
    return (1.0 - x) ** 2 + 100.0 * (y - x * x) ** 2


def test_quadratic_bowl() -> None:
    res = nelder_mead(lambda p: (p[0] - 2.0) ** 2 + (p[1] + 1.0) ** 2, [0.0, 0.0])
    assert res.converged
    assert res.x[0] == pytest.approx(2.0, abs=1e-6)
    assert res.x[1] == pytest.approx(-1.0, abs=1e-6)
    assert res.fx == pytest.approx(0.0, abs=1e-12)


def test_rosenbrock_vs_scipy() -> None:
    ours = nelder_mead(_rosenbrock, [-1.2, 1.0], maxiter=10_000)
    ref = sopt.minimize(
        _rosenbrock, [-1.2, 1.0], method="Nelder-Mead",
        options={"xatol": 1e-10, "fatol": 1e-14, "maxiter": 10_000},
    )
    assert ours.converged
    assert ours.x[0] == pytest.approx(1.0, abs=1e-5)
    assert ours.x[1] == pytest.approx(1.0, abs=1e-5)
    assert ours.fx == pytest.approx(float(ref.fun), abs=1e-9)


def test_non_convergence_reported() -> None:
    res = nelder_mead(_rosenbrock, [-1.2, 1.0], maxiter=3)
    assert not res.converged
    assert res.iterations == 3


def test_input_validation() -> None:
    with pytest.raises(ValueError):
        nelder_mead(lambda p: 0.0, [])
    with pytest.raises(ValueError):
        nelder_mead(lambda p: p[0], [float("nan")])
